### ISSUE TITLE
How should `Datasource` handle data and `overwrite=True`

### ISSUE THREAD

**Issue by gareth-j** [openghg/openghg#591]
How do we want `Datasource` to handle new data being added? Do we always want data to be combined when new data is added?

**Comment by rt17603** [comment]
Scenario 1
 - New data added
   - Add new data source
   - version 1 added

Scenario 2
 - New data, non-overlapping date range
   - Combine with previous data source (check overlap)
   - new version assigned, version 2

Scenario 3a
 - New data, overlapping date range / replacement date range
   - [Currently: Add data check for overlapping, try combining and drop duplicates]
   - Don't add

Scenario 3b
 - New data, overlapping date range / replacement date range + overwrite
  - "Retire previous data to previous data"

Scenario 3c
 - New data, overlapping date range / replacement date range + overwrite + update
   - "Replace overlapping date range" - delete old, insert new

**Comment by rt17603** [comment]
Caveats: need to make sure NaN values and data not being present is treated differently. NaN should be replaced and then dropped rather than just ignored.

**Comment by gareth-j** [comment]
I'm going to link this with #319 as the ability to pass in these flags will be helped by a rework of `assign_data`.

**Comment by rt17603** [comment]
Alternatively, could set this up as `mode` option inputs e.g.
 - mode = None - default
 - mode = "overwrite" - overwrite description above
 - mode = "update" - update description above

IF overwrite / update are essentially mutually exclusive

**Comment by rt17603** [comment]
Check and close #465 as appropriate in relation to this.

**Comment by rt17603** [comment]
@joe-pitt - if you could take a look at this workflow let me know if this does/doesn't address the question you had around updating data in the object store.

**Comment by joe-pitt** [comment]
Yes I think on_overlap input = "replace" will be useful for data that gets regularly updated (e.g. NRT data). I guess we can decide between us at some later point exactly which data should be "retired" (with a version number) and which should be replaced, but it would be good to have the functionality to do both.

**Comment by rt17603** [comment]
I'm considering what should be done for Scenario 2 (Previous data source, new data added, non-overlapping date range) when the flags are passed and I think this may be better as an `if_exists` input rather than `on_overlap` so something like:
 - `if_exists = "new"` always create a new version
 - `if_exists = "replace"` replace any overlapping existing data and insert into current version

**Comment by rt17603** [comment]
From discussion on 05/06/2023 with Joe Pitt:

Better to consider this as two separate questions:
1. How do we want to combine the new data with any current data?
2. How do we want to handle saving / deleting any current data (e.g. new version)?

Proposal:
 - To address (1) - `if_exists` - ties to None, "new" and "replace" as details of how to act when combining new and current data.
 - To address (2) - `save_current` - ties to whether to save the current data *in it's exact form* and create a new version for the new / combined data.

Default behaviour:
 - `if_exists=None`, `save_current=None` - only add new data which doesn't overlap with current, same version (`new_version=False`)
 - `if_exists="new"` / `"replace"`, `save_current=None` - save current data as previous version (`new_version=True`)
 - `if_exists="new"` / `"replace"`, `save_current=True` / `False` - use inputs for these keywords

**Comment by rt17603** [comment]
Applying these changes in steps and merging into `'new_object_store'` branch before merging into `'devel'` as this may be a breaking change for those running against a current object store. Plan is to create new object store using this parallel branch first before merging into `'devel'`

- [x] 1. Apply updates to current default workflow (small tweaks to overwrite actions as needed) - PR #653 
- [ ] 2. Create `if_exists` and `save_current` keywords. Implement "new" case. Depreceate overwrite flag
- [ ] 3. Update "replace"/"insert" case for `if_exists` functionality

**Comment by rt17603** [comment]
Additionally want to address case where we're trying to add identical data (hashes are the same) - include `force` flag to cover this specific case.
